packages/protocol/contracts/tokenvault/ERC721Vault.sol

```diff
@@ -152,34 +135,19 @@ contract ERC721Vault is BaseNFTVault, IERC721ReceiverUpgradeable {
             revert VAULT_INVALID_SRC_CHAIN_ID();
         }
 
-        (CanonicalNFT memory nft,,, uint256[] memory tokenIds) =
+        (CanonicalNFT memory ctoken,,, uint256[] memory tokenIds) =
             abi.decode(message.data[4:], (CanonicalNFT, address, address, uint256[]));
 
-        if (nft.addr == address(0)) revert VAULT_INVALID_TOKEN();
+        if (ctoken.addr == address(0)) revert VAULT_INVALID_TOKEN();
 
-        unchecked {
-            if (bridgedToCanonical[nft.addr].addr != address(0)) {
-                for (uint256 i; i < tokenIds.length; ++i) {
-                    BridgedERC721(nft.addr).mint(message.owner, tokenIds[i]);
-                }
-            } else {
-                for (uint256 i; i < tokenIds.length; ++i) {
-                    ERC721Upgradeable(nft.addr).safeTransferFrom({
-                        from: address(this),
-                        to: message.owner,
-                        tokenId: tokenIds[i]
-                    });
-                }
-            }
-        }
-
-        // send back Ether
+        // Transfer the ETH and tokens back to the owner
+        _transferTokens(ctoken, message.owner, tokenIds);
         message.owner.sendEther(message.value);
 
         emit TokenReleased({
             msgHash: msgHash,
             from: message.owner,
-            token: nft.addr,
+            token: ctoken.addr,
```

Comment:
Same as above ? Emit address returned by `_transferTokens(ctoken, message.owner, tokenIds);`) ?

Reply:
Maybe we emit both the ctoken address and the actual token address?

Reply:
Dont think it is necessary as it might not even exists (bridged). recall is always called on the source chain, so from TokenReleased (msgHash is from the same source chain) event we could identify the transactions.

Reply:
If bridged token doesn't know exist on the local chain, address(0) can be used.

Reply:
`TokenSent`, `TokenReceived`, and `TokenReleased` now all log the actual token address. I don't know what's convenient for offchain applications so can't really judge what data should be logged, but having those three events log the token addresses in a consistent way seems to make sense to me. So if the canonical address is something that would be helpful, I would add it to all three (or if not, to none).

Reply:
@Brechtpd I do think canonical address shall also be logged which will be convenient for the off-chain apps who doesn't keep track of all mapping between canonical token and the actual token, given that the extra cost is really small.